File: src/lib/types/workflow.ts
/**
 * Workflow management types for dynamic workflow system
 */

export interface AgentRegistry {
  id: string;
  name: string;
  module_path: string;
  factory_function: string;
  description: string;
  config_schema: Record<string, any>;
  tools: string[];
  created_at: string;
  updated_at: string;
}

export interface ToolRegistry {
  id: string;
  name: string;
  module_path: string;
  function_name: string;
  description: string;
  input_schema: Record<string, any>;
  output_schema: Record<string, any>;
  created_at: string;
  description_version: string;
}

export interface WorkflowDefinition {
  id: string;
  name: string;
  description: string;
  current_revision_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowRevision {
  id: string;
  workflow_definition_id: string;
  version: number;
  is_published: boolean;
  revision_metadata: {
    author?: string;
    notes?: string;
    changelog?: string;
  };
  created_at: string;
  published_at: string | null;
}

export interface WorkflowStep {
  id: string;
  revision_id: string;
  step_order: number;
  step_name: string;
  agent_id: string | null;
  execution_mode: 'sequential' | 'concurrent' | 'conditional';
  input_mapping: Record<string, any>;
  output_mapping: Record<string, any>;
  checkpoint_enabled: boolean;
  retry_policy: {
    max_retries?: number;
    backoff?: string;
  } | null;
  step_config: Record<string, any>;
}

export interface WorkflowInstance {
  id: string;
  job_id: string;
  revision_id: string;
  current_step_id: string | null;
  instance_state: Record<string, any>;
  started_at: string | null;
  completed_at: string | null;
  status: 'running' | 'paused' | 'completed' | 'failed';
}

export interface WorkflowWithSteps {
  definition: WorkflowDefinition;
  revision: WorkflowRevision;
  steps: WorkflowStep[];
}

export interface CreateWorkflowRevisionRequest {
  workflow_definition_id: string;
  revision_metadata: {
    author: string;
    notes: string;
    changelog?: string;
  };
  steps: Omit<WorkflowStep, 'id' | 'revision_id'>[];
}

export interface PublishRevisionRequest {
  revision_id: string;
}

export interface ValidationResult {
  is_valid: boolean;
  errors: string[];
  warnings: string[];
}

File: backend/api/schemas/workflow.py
"""Pydantic schemas for workflow administration APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class WorkflowDefinitionCreate(BaseModel):
    name: str = Field(..., max_length=255)
    description: Optional[str] = None


class WorkflowDefinitionUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = None


class WorkflowDefinitionOut(BaseModel):
    id: UUID
    name: str
    description: Optional[str]
    current_revision_id: Optional[UUID]
    current_version: Optional[int]
    created_at: datetime
    updated_at: datetime


class WorkflowStepCreate(BaseModel):
    step_order: int
    step_name: str = Field(..., max_length=255)
    agent_id: Optional[UUID] = None
    execution_mode: str
    input_mapping: Optional[Dict[str, Any]] = None
    output_mapping: Optional[Dict[str, Any]] = None
    checkpoint_enabled: bool = True
    retry_policy: Optional[Dict[str, Any]] = None
    step_config: Optional[Dict[str, Any]] = None


class WorkflowRevisionCreate(BaseModel):
    revision_metadata: Optional[Dict[str, Any]] = None
    steps: List[WorkflowStepCreate]
    publish: bool = False


class WorkflowStepOut(BaseModel):
    id: UUID
    step_order: int
    step_name: str
    execution_mode: str
    agent_id: Optional[UUID]
    checkpoint_enabled: bool
    input_mapping: Optional[Dict[str, Any]]
    output_mapping: Optional[Dict[str, Any]]
    retry_policy: Optional[Dict[str, Any]]
    step_config: Optional[Dict[str, Any]]


class WorkflowRevisionOut(BaseModel):
    id: UUID
    version: int
    is_published: bool
    revision_metadata: Optional[Dict[str, Any]]
    created_at: datetime
    published_at: Optional[datetime]
    steps: List[WorkflowStepOut]


class RevisionValidationResult(BaseModel):
    is_valid: bool
    errors: List[str] = []
